[DX-815] Plugins Overview Refactor (#5330)
TykTechnologies/tyk-docs: tyk-docs/content/product-stack/tyk-gateway/advanced-configurations/plugins/api-config/oas.md

---
date: 2024-06-25T12:59:42Z
title: Configuring Plugins for Tyk OAS APIs
description: "This section explains how to configure Tyk OAS APIs to use plugin bundles deployed on a remote web server"
tags: [ "Tyk OAS plugins" ]
---

> An API can be configured so that one or more of its associated plugins can execute at different phases of the request life cycle. Each plugin configuration serves to identify the plugin source file path and the name of the corresponding function, triggered at each request lifecycle stage.

Author: An API can be configured so that one or more of its associated plugins can execute at different phases of the request / response life cycle. Each plugin configuration serves to identify the plugin source file path and the name of the corresponding function, triggered at each request / response lifecycle stage.

      "name": "OAS APIs Plugins",
      "state": {
        "active": true,
        "internal": false
      }
    },
    "middleware": {
      "global": {
        "pluginConfig": {
          "driver": "goplugin"
        },
        "postAuthenticationPlugins": [
          {
            "enabled": true,
            "functionName": "post_authentication_func",
            "path": "/path/to/plugin1.so",
            "rawBodyOnly": true,
            "requireSession": true
          }
        ],
        "postPlugins": [
          {
            "enabled": true,
            "functionName": "postplugin",
            "path": "/path/to/plugin1.so",
            "rawBodyOnly": true,
            "requireSession": true
          }
        ],
        "prePlugins": [
          {
            "enabled": true,
            "functionName": "pre-plugin",
            "path": "/path/to/plugin1.so"
          }
        ],
        "responsePlugins": [
          {
            "enabled": true,
            "functionName": "Response",
            "path": "/path/to/plugin1.so",
            "rawBodyOnly": true,
            "requireSession": true
          }
        ]
      }
    }
  }
}
```

> We can see from the example above that the `global` object within the `middleware` section of the `x-tyk-api-gateway` configuration is used to configure plugins in a Tyk OAS API. The `pluginConfig` section contains the `driver` parameter that is used to configure the plugin implementation [language]({{< ref "/plugins/supported-languages#plugin-driver-names" >}}).

Author: updated

A list of plugins can be configured for each plugin type, e.g. Post Authentication (`postAuthenticationPlugins`), Pre (`prePlugins`) etc. In the example above we can see that there are Golang post authentication, post, pre and response plugins configured.

Each plugin has the following configuration parameters:

- `enabled`: When true, enables the plugin.
- `functionName`: The name of the function that implements the plugin within the source file.
- `path`: The path to the plugin source file. 
- `rawBodyOnly`: When true, indicates that only the raw body should be processed.
- `requireSession`: When true, indicates that the plugin requires an active session. This is applicable only for Post, Post Authentication and Response plugins.

---

## Configuring plugins in the API Designer {#tyk-oas-dashboard}

Click on the APIs menu item in the *API Management* menu of Dashboard and select your OAS API to display the OAS API editor screen. Subsequently, follow the steps below:

> #### Step 1: Configure implementation language for your plugin

Reviewer: ```suggestion
#### Step 1: Configure plugin type and custom data
```

> Scroll down until the *Plugins Configuration* section is displayed with the option to configure a plugin driver and a list of plugin types, e.g. *Pre Plugin*, *Post Plugin* etc. 

Reviewer: ```suggestion
In the *Plugins Configuration* section, select the *Plugin Driver*, which tells Tyk which type of plugin to expect: Go, gRPC, JavaScript (OTTO), Lua or Python.

You can configure custom data that will be made available to your plugin function as a JSON formatted object in the *Config Data* option.
```

{{< img src="/img/plugins/plugins_oas_api_driver_options.png" alt="OAS API Plugins Driver Config" >}}

> Select the implementation language of your plugins.

Reviewer: ```suggestion
```

> #### Step 2: Add plugins

Reviewer: ```suggestion
#### Step 2: Configure the custom plugins
```

> For each type of plugin to configure, click on the *Add Plugin* button to display a plugin configuration section:

Reviewer: ```suggestion
For each plugin that you wish to register with the API, click on the **Add Plugin** button to display a plugin configuration section:
```

{{< img src="/img/plugins/plugins_oas_api_source_config.png" alt="OAS Plugins Config Section" >}}

Complete the following fields:

> - `Function Name`: Enter the function name that implements the required behavior for your plugin.

Reviewer: ```suggestion
- `Function Name`: Enter the name of the function within your plugin code that Tyk should invoke.
```

- `Path`: Enter the path to the source file that contains the function that implements your plugin.

> - `Raw Body Only`: Optionally, toggle the *Raw Body Only* switch to true when you do not wish to fill body in request or response object for your plugins.

Reviewer: ```suggestion
- `Raw Body Only`: toggle this to true when you do not wish to fill body in request or response object for your plugins.
```

Not sure about this, doesn't really make sense - checking

Author: Thanks Andy I have created a [ticket](https://tyktech.atlassian.net/browse/DX-1658) to update in another iteration.

@letzya
